privacyidea.py

```diff
@@ -2,131 +2,241 @@
+        fm.setKeepMessages()
+
+        if not self.pi:
+            fm.add(FacesMessage.SEVERITY_ERROR, "The script could not be set up properly, please check the log files")
+            return False
+
+        authenticationService = CdiUtil.bean(AuthenticationService)
+        identity = CdiUtil.bean(Identity)
+
         if step == 1:
-            print "privacyIDEA. Authenticate against privacyIDEA"
-            credentials = self.identity.getCredentials()
+            credentials = identity.getCredentials()
             user_name = credentials.getUsername()
             user_password = credentials.getPassword()
 
-            # TODO: Run /validate/check for the first time with user_name and user_password
-            if user_password == "pin":
-                print("privacyIDEA. The first auth step was successful")
-                # If the user is successfully authenticated, we need to add "aut_user" to the session.
-                sessionId = self.sessionIdService.getSessionId() # fetch from persistence
-                sessionId.getSessionAttributes().put("pi_transaction_id", "blabla")
-                self.sessionIdService.updateSessionId(sessionId)
-                print("privacyIDEA. Updating pi_transaction_id")
-                self.authenticationService.authenticate(user_name)
-
-                print("privacyIDEA. Authenticate. Logged in: {0!s}".format(user_name))
-                # We have to return True to get to the next step!
-                return True
-
-        elif step == 2:
-            # TODO: Run /validate/check a second time and respond the challenge
-            print("privacyIDEA. Challenge Response")
-            credentials = self.identity.getCredentials()
-            #user_name = credentials.getUsername()
-            sessionId = self.sessionIdService.getSessionId()
-            user_name = sessionId.getSessionAttributes().get("auth_user")
-            user_password = credentials.getPassword()
-            print("privacyIDEA. Authenticate. CHALLENGE RESPONSE!")
-            print("privacyIDEA. {0!s} / {1!s}.".format(user_name, user_password))
-            # TODO: Send the password and the transaction_id to privacyIDEA.
-            # Pass the user into the session.
-            if user_password == "otp":
-                return True
+            if user_name:
+                self.addToSession("currentUser", user_name)
+
+                response = self.pi.validateCheck(user_name, user_password)
+                if response:
+                    # First check if what was entered is sufficient to log in
+                    if response.getValue():
+                        logged_in = authenticationService.authenticate(user_name)
+                        if logged_in:
+                            self.addToSession("auth_success", True)
+                        return logged_in
+
+                    # If not, check if transaction was triggered
+                    if response.getTransactionID():
```

Comment:
I personally usually do not like early exits, since it could change the flow, if you do not take care.

So I recommend to do an elif, here:

~~~~python
    elif response.getTransactionID():
~~~~
which makes sure, that you really only enter this, if the response.getValue() was not True.